nexus/metrics-producer-gc/src/lib.rs

```diff
@@ -0,0 +1,369 @@
+        // producer's `time_modified` should prune our one producer.
+        let pruned = prune_expired_producers(
+            &opctx,
+            &datastore,
+            producer_time_modified + Duration::from_secs(1),
+        )
+        .await
+        .expect("failed to prune expired producers");
+        let expected_success =
+            [producer.id()].into_iter().collect::<BTreeSet<_>>();
+        assert_eq!(pruned.successes, expected_success);
+        assert!(pruned.failures.is_empty());
+
+        // GC'ing again with the same expiration should do nothing, because we
+        // already pruned the producer.
+        let pruned = prune_expired_producers(
+            &opctx,
+            &datastore,
+            producer_time_modified + Duration::from_secs(1),
+        )
+        .await
+        .expect("failed to prune expired producers");
+        assert!(pruned.successes.is_empty());
+        assert!(pruned.failures.is_empty());
+
+        db.cleanup().await.unwrap();
+        logctx.cleanup_successful();
+    }
+
+    #[tokio::test]
+    async fn test_prune_expired_producers_notifies_collector() {
+        // Setup
+        let logctx = dev::test_setup_log(
+            "test_prune_expired_producers_notifies_collector",
+        );
+        let mut db = test_setup_database(&logctx.log).await;
+        let (opctx, datastore) =
+            datastore_test(&logctx, &db, Uuid::new_v4()).await;
+
+        let mut collector = httptest::Server::run();
+
+        // Insert an Oximeter collector
+        let collector_info = OximeterInfo::new(&params::OximeterInfo {
+            collector_id: Uuid::new_v4(),
+            address: collector.addr(),
+        });
+        datastore
+            .oximeter_create(&collector_info)
+            .await
+            .expect("failed to insert collector");
+
+        // Insert a producer.
+        let producer = ProducerEndpoint::new(
+            &nexus::ProducerEndpoint {
+                id: Uuid::new_v4(),
+                kind: nexus::ProducerKind::Service,
+                address: "[::1]:0".parse().unwrap(), // unused
+                base_route: "/".to_string(),         // unused
+                interval: Duration::from_secs(0),    // unused
+            },
+            collector_info.id,
+        );
+        datastore
+            .producer_endpoint_create(&producer)
+            .await
+            .expect("failed to insert producer");
+
+        let producer_time_modified =
+            read_time_modified(&datastore, producer.id()).await;
+
+        // GC'ing expired producers with an expiration time _newer_ than our
+        // producer's `time_modified` should prune our one producer and notify
+        // the collector that it's doing so.
+        collector.expect(
+            Expectation::matching(request::method_path(
+                "DELETE",
+                format!("/producers/{}", producer.id()),
+            ))
+            .respond_with(status_code(404)),
```

Comment:
I was surprised to see a 404 here. But it looks like the pruning method only logs the error, right?

Reply:
Ah, I think this is leftover from when notifying the collector was more load bearing. If we try to delete a nonexistent producer, oximeter will return a 404 (reasonable), so I was testing that we succeeded whether we got a 204 or a 404, and then kept the wrong one when I was cleaning up. I'll switch this to 204.

You are correct that we only log an error - it's a one shot / best-effort optimization.